File: lib/utils/patient_search_index.dart
String normalizePatientSearchText(String value) {
  return value.trim().replaceAll(RegExp(r'\s+'), ' ').toLowerCase();
}

List<String> buildPatientSearchPrefixes(String patientName) {
  final normalizedName = normalizePatientSearchText(patientName);
  if (normalizedName.isEmpty) {
    return const <String>[];
  }

  const spaceRune = 0x20;
  final runes = normalizedName.runes.toList(growable: false);
  final componentStarts = <int>[0];
  for (var index = 0; index < runes.length; index++) {
    if (runes[index] == spaceRune) {
      componentStarts.add(index + 1);
    }
  }

  final prefixes = <String>{};
  for (final start in componentStarts) {
    for (var end = start + 1; end <= runes.length; end++) {
      if (runes[end - 1] != spaceRune) {
        prefixes.add(String.fromCharCodes(runes.sublist(start, end)));
      }
    }
  }

  return List<String>.unmodifiable(prefixes);
}

File: firebase/scripts/patient_search_index.mjs
export function normalizePatientSearchText(value) {
  return value.trim().replace(/\s+/gu, ' ').toLowerCase();
}

export function buildPatientSearchPrefixes(patientName) {
  const normalizedName = normalizePatientSearchText(patientName);
  if (!normalizedName) return [];

  const codePoints = [...normalizedName];
  const componentStarts = [0];
  for (let index = 0; index < codePoints.length; index += 1) {
    if (codePoints[index] === ' ') {
      componentStarts.push(index + 1);
    }
  }

  const prefixes = new Set();
  for (const start of componentStarts) {
    for (let end = start + 1; end <= codePoints.length; end += 1) {
      if (codePoints[end - 1] !== ' ') {
        prefixes.add(codePoints.slice(start, end).join(''));
      }
    }
  }
  return [...prefixes];
}
